firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/FirebaseAppDistribution.java

```diff
@@ -296,27 +324,71 @@ AppDistributionReleaseInternal getCachedLatestRelease() {
         AlertDialog.BUTTON_POSITIVE,
         context.getString(R.string.update_yes_button),
         (dialogInterface, i) ->
-            // show download progress in notification manager
+        // show download progress in notification manager
+        {
+          synchronized (updateTaskLock) {
             updateApp(true)
-                .addOnSuccessListener(unused -> updateAlertDialogTask.setResult(null))
-                .addOnFailureListener(updateAlertDialogTask::setException));
+                .addOnProgressListener(progress -> cachedUpdateIfNewReleaseUpdateProgress(progress))
+                .addOnSuccessListener(unused -> cachedUpdateIfNewReleaseTask.setResult())
+                .addOnFailureListener(cachedUpdateIfNewReleaseTask::setException);
+          }
+        });
 
     alertDialog.setButton(
         AlertDialog.BUTTON_NEGATIVE,
         context.getString(R.string.update_no_button),
         (dialogInterface, i) -> {
           dialogInterface.dismiss();
-          updateAlertDialogTask.setException(
-              new FirebaseAppDistributionException(
-                  Constants.ErrorMessages.UPDATE_CANCELED,
-                  FirebaseAppDistributionException.Status.INSTALLATION_CANCELED));
+          synchronized (updateTaskLock) {
+            cachedUpdateIfNewReleaseUpdateProgress(
+                UpdateProgress.builder()
+                    .setApkFileTotalBytes(UNKNOWN_RELEASE_FILE_SIZE)
+                    .setApkBytesDownloaded(UNKNOWN_RELEASE_FILE_SIZE)
+                    .setUpdateStatus(UpdateStatus.UPDATE_CANCELED)
+                    .build());
+            setCachedUpdateIfNewReleaseCompletionError(
+                new FirebaseAppDistributionException(
+                    Constants.ErrorMessages.UPDATE_CANCELED,
+                    FirebaseAppDistributionException.Status.INSTALLATION_CANCELED));
+          }
         });
 
     alertDialog.show();
-    return updateAlertDialogTask.getTask();
+    synchronized (updateTaskLock) {
+      return cachedUpdateIfNewReleaseTask;
+    }
   }
 
   void setInstallationResult(int resultCode) {
     this.updateAppClient.setInstallationResult(resultCode);
   }
+
+  private void setCachedUpdateIfNewReleaseCompletionError(FirebaseAppDistributionException e) {
+    synchronized (updateTaskLock) {
+      if (cachedUpdateIfNewReleaseTask != null && !cachedUpdateIfNewReleaseTask.isComplete()) {
+        cachedUpdateIfNewReleaseTask.setException(e);
+      }
+    }
+  }
+
+  private void setCachedUpdateIfNewReleaseCompletionError(
+      Exception e, FirebaseAppDistributionException defaultFirebaseException) {
+    if (e instanceof FirebaseAppDistributionException) {
+      setCachedUpdateIfNewReleaseCompletionError((FirebaseAppDistributionException) e);
+    } else {
+      setCachedUpdateIfNewReleaseCompletionError(defaultFirebaseException);
+    }
+  }
+
+  private void cachedUpdateIfNewReleaseUpdateProgress(UpdateProgress progress) {
```

Comment:
nit: Java methods often start with verbs. Maybe this could be something like `postProgressToCachedUpdateIfNewReleaseTask` ?